File: servertier/src/main/scala/sgad/servertier/dataaccess/data/userdata/UnitPossession.scala
/**
 * FILE: UnitPossession.scala
 * PERCORSO /Codice/sgad/servertier/src/main/scala/sgad/servertier/dataaccess/data/userdata
 * DATA CREAZIONE: 19 Febbraio 2014
 * AUTORE: ProTech
 * EMAIL: protech.unipd@gmail.com
 *
 * Questo file è proprietà del gruppo ProTech, viene rilasciato sotto licenza Apache v2.
 *
 * DIARIO DELLE MODIFICHE:
 * 2014-02-19 - Creazione della classe - Segantin Fabio
 */
package sgad.servertier.dataaccess.data.userdata

import sgad.servertier.dataaccess.data.shareddata.`Unit`

/**
 * Classe per la rappresentazione di un'unità con associata una quantità.
 * @constructor
 * @param quantity La quantità di unità.
 * @param unit Il tipo di unità.
 */
class UnitPossession(private var quantity: Int, private val unit: `Unit`) {
	if (unit == null
		|| quantity < 0)
		throw new IllegalArgumentException

	/**
	 * Metodo getter per l'attributo quantity.
	 * @return La quantità di unità.
	 */
	def getQuantity = quantity

	/**
	 * Metodo getter per l'attributo unit.
	 * @return Il tipo di unità.
	 */
	def getUnit = unit

	/**
	 * Metodo setter per l'attributo quantity.
	 * @param quantity La quantità di unità, deve essere >=0.
	 */
	def setQuantity(quantity: Int) = {
		if (quantity < 0)
			throw new IllegalArgumentException
		this.quantity = quantity
	}

	/**
	 * Metodo di confronto
	 * @param other L'oggetto con cui fare il confronto.
	 * @return True or false in base all'esito.
	 */
	override def equals(other: Any): Boolean = {
		if (!other.isInstanceOf[UnitPossession])
			false
		else
			(unit == other.asInstanceOf[UnitPossession].unit
				&& quantity == other.asInstanceOf[UnitPossession].quantity)
	}

	/**
	 * Metodo getter per la key di UnitPossession.
	 * @return La chiave relativa all'unità.
	 */
	def getKey = unit.getKey
}

File: Codice/clienttier/model/personaldata/UnitPossession.js
/**
 * FILE: UnitPossession.js
 * PERCORSO /Codice/sgad/clienttier/model/personaldata/UnitPossession.js
 * DATA CREAZIONE: 18 Febbraio 2014
 * AUTORE: ProTech
 * EMAIL: protech.unipd@gmail.com
 *
 * Questo file è proprietà del gruppo ProTech, viene rilasciato sotto licenza Apache v2.
 *
 * DIARIO DELLE MODIFICHE:
 * 2014-02-20 - Implementazione dei metodi getQuantity(), setQuantity(quantity), getUnit(), getKey() e valueOf() - Gatto Francesco
 * 2014-02-18 - Creazione della classe - Gatto Francesco
 */

//Eredita da Observable.
UnitPossession.prototype = new Observable();

//Costruttore della classe.
UnitPossession.prototype.constructor = UnitPossession;
/**
 * Classe per la rappresentazione di un'unità con associata una quantità.
 * @extends Observable
 * @constructor
 * @param {int} quantity La quantità di unità.
 * @param {Unit} unit Il tipo di unità.
 **/
function UnitPossession(quantity, unit) {
	/**
	 * La quantità di unità.
	 * @type {int}
	 * @private
	 */
	this.quantity = quantity;
	/**
	 * Il tipo di unità.
	 * @type {Unit}
	 * @private
	 */
	this.unit = unit;
}
/**
 * Metodo getter per la quantità di unità.
 * @return {int} La quantità di unità.
 */
UnitPossession.prototype.getQuantity = function() {
    return this.quantity;
};

/**
 * Metodo setter per la quantità di unità.
 * @param {int} quantity La quantità da impostare.
 * @return {void}
 */
UnitPossession.prototype.setQuantity = function(quantity) {
    this.quantity = quantity;
	this.notify();
};

/**
 * Fornisce un valore numerico rappresentante una quantità.
 * @override
 * @return {int} Il numero di unità.
 */
UnitPossession.prototype.getState = function() {
	return this.getQuantity();
};

/**
 * Imposta il valore numerico rappresentante la quantità.
 * @param {int} quantity La quantità da impostare.
 * @return {void}
 */
UnitPossession.prototype.setState = function(quantity) {
	this.quantity = quantity;
};

/**
 * Metodo getter per il tipo di unità.
 * @return {Unit} il tipo di unità.
 */
UnitPossession.prototype.getUnit = function() {
    return this.unit;
};

/**
 * Metodo getter per la chiave.
 * @return {string} La chiave.
 */
UnitPossession.prototype.getKey = function() {
	return this.unit.getKey();
};

/**
 * Ridefinizione del metodo valueOf di UnitPossession.
 * @override
 * @return {string} L'identificativo di UnitPossession.
 */
UnitPossession.prototype.valueOf = function() {
    return  JSON.stringify({unit : this.getUnit(), quantity : this.getQuantity()});
};